File: Ruby/kmeans_clustering.rb
require 'matrix'

def k_means(points, k, max_iters=100)
  # Randomly initialize centroids
  centroids = points.sample(k)

  max_iters.times do
    # Assign each point to the nearest centroid
    labels = points.map do |point|
      centroids.each_with_index.min_by { |c, _| (Vector.elements(point) - Vector.elements(c)).magnitude }[1]
    end

    # Recalculate centroids as the mean of assigned points
    new_centroids = Array.new(k) { |i| points.select.with_index { |_, j| labels[j] == i }.transpose.map { |dim| dim.sum / dim.size } }

    # Check for convergence
    break if centroids == new_centroids
    centroids = new_centroids
  end

  [centroids, labels]
end

File: R/kmeans_clustering.R
k_means <- function(X, k, max_iters=100) {
  # Perform K-means clustering on a dataset X
  #
  # Parameters:
  # X: Dataset with samples in rows
  # k: Number of clusters
  # max_iters: Maximum number of iterations
  
  # Randomly initialize centroids
  set.seed(42)
  centroids <- X[sample(1:nrow(X), k),]
  
  for (iter in 1:max_iters) {
    # Assign each data point to the nearest centroid
    labels <- apply(X, 1, function(x) {
      which.min(sapply(1:k, function(i) sum((x - centroids[i,])^2)))
    })
    
    # Recalculate centroids as the mean of assigned points
    new_centroids <- sapply(1:k, function(i) colMeans(X[labels == i, , drop=FALSE]))
    
    # Check for convergence
    if (all(centroids == t(new_centroids))) break
    centroids <- t(new_centroids)
  }
  
  list(centroids = centroids, labels = labels)
}
